File: src/lib/github/access.ts
import * as github from '@pulumi/github';
import { Output } from '@pulumi/pulumi';

import {
  RepositoryConfig,
  RepositoryTeamConfig,
} from '../../model/config/repository';
import { StringMap } from '../../model/map';
import { githubHandle } from '../configuration';

/**
 * Creates all access permissions for a repository.
 *
 * @param {RepositoryConfig} repository the repository configuration
 * @param {github.Repository} githubRepository the GitHub repository
 * @param {StringMap<github.Team>} githubTeams the GitHub teams
 */
export const createRepositoryAccess = (
  repository: RepositoryConfig,
  githubRepository: github.Repository,
  githubTeams: StringMap<github.Team>,
) => {
  new github.RepositoryCollaborator(
    `github-repository-admin-${repository.name}`,
    {
      repository: githubRepository.name,
      username: githubHandle,
      permission: 'admin',
    },
    {
      dependsOn: [githubRepository],
      retainOnDelete: true,
    },
  );

  repository.teams.forEach(async (team) => {
    createTeamAccess(githubRepository, team, githubTeams[team.name]?.id);
  });
};

/**
 * Creates access permissions for the team to the repository.
 *
 * @param {github.Repository} repository the repository
 * @param {RepositoryTeamConfig} team the team configuration
 * @param {number} teamId the team id
 */
const createTeamAccess = (
  repository: github.Repository,
  team: RepositoryTeamConfig,
  teamId: Output<string>,
) =>
  repository.name.apply(
    (repositoryName) =>
      new github.TeamRepository(
        `github-team-repository-${repositoryName}-${team.name}`,
        {
          repository: repositoryName,
          teamId: teamId,
          permission: repositoryRoleToGitHubRole(team.role),
        },
        {
          dependsOn: [repository],
          retainOnDelete: true,
        },
      ),
  );

/**
 * Maps a repository role to a Harbor role.
 *
 * @param role the repository role
 * @returns the corresponding Harbor role
 */
const repositoryRoleToGitHubRole = (role: string): string => {
  switch (role) {
    case 'developer':
      return 'maintain';
    default:
      return '';
  }
};

File: pkg/lib/github/repositories/access.go
package repositories

import (
	"fmt"

	"github.com/pulumi/pulumi-github/sdk/v6/go/github"
	"github.com/pulumi/pulumi/sdk/v3/go/pulumi"

	"github.com/muhlba91/fh-burgenland-bswe-assignment-infrastructure/pkg/lib/config"
	"github.com/muhlba91/fh-burgenland-bswe-assignment-infrastructure/pkg/model/config/repository"
)

// createAccess creates access permissions for the given repository based on the provided configuration.
// ctx: The Pulumi context for resource creation.
// repository: The configuration for the repository.
// repo: The Pulumi GitHub repository resource.
// githubTeams: A map of GitHub teams for potential repository team associations.
func createAccess(
	ctx *pulumi.Context,
	repository *repository.Config,
	repo *github.Repository,
	githubTeams map[string]*github.Team,
) error {
	_, collErr := github.NewRepositoryCollaborator(
		ctx,
		fmt.Sprintf("github-repository-admin-%s", repository.Name),
		&github.RepositoryCollaboratorArgs{
			Repository: repo.Name,
			Username:   pulumi.String(config.OwnerHandle),
			Permission: pulumi.String("admin"),
		},
	)
	if collErr != nil {
		return collErr
	}

	for _, team := range repository.Teams {
		role := repositoryRoleToGitHubRole(team.Role)
		if len(role) == 0 {
			continue
		}

		ghTeam, exists := githubTeams[team.Name]
		if !exists {
			return fmt.Errorf("team %s not found for repository %s", team.Name, repository.Name)
		}

		_ = repo.Name.ApplyT(func(name string) error {
			_, tcErr := github.NewTeamRepository(
				ctx,
				fmt.Sprintf("github-team-repository-%s-%s", name, team.Name),
				&github.TeamRepositoryArgs{
					Repository: repo.Name,
					TeamId:     ghTeam.ID(),
					Permission: pulumi.String(role),
				},
				pulumi.DependsOn([]pulumi.Resource{repo}),
				pulumi.RetainOnDelete(true),
			)
			return tcErr
		})
	}

	return nil
}

// repositoryRoleToGitHubRole maps a custom repository role to a GitHub permission string.
// role: The custom repository role.
func repositoryRoleToGitHubRole(role string) string {
	switch role {
	case "owner":
		return "maintain"
	case "developer":
		return "maintain"
	default:
		return ""
	}
}
